### ISSUE TITLE
userns: docker build creates directories with wrong permissions

### ISSUE THREAD

**Issue by skohanim** [docker/docker#20395]
**Description of problem:**
When using `--userns-remap=default` docker build COPY creates non existing directory hierarchy with wrong owner.
This doesn't happen without userns remapping.

**docker version:**
```
Client:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.1
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   9e83765
 Built:        Thu Feb 11 19:27:08 2016
 OS/Arch:      linux/amd64
```

**docker info:**
```
Containers: 19
 Running: 3
 Paused: 0
 Stopped: 16
Images: 26
Server Version: 1.10.1
Storage Driver: devicemapper
 Pool Name: docker-8:1-135103-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: ext4
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 1.431 GB
 Data Space Total: 107.4 GB
 Data Space Available: 38.34 GB
 Metadata Space Used: 3.805 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.144 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/231072.231072/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/231072.231072/devicemapper/devicemapper/metadata
 Library Version: 1.02.77 (2012-10-15)
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 3.13.0-67-generic
Operating System: Ubuntu 14.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.955 GiB
Name: vagrant-ubuntu-trusty-64
ID: WZZC:Y7VT:T5JN:GALQ:JJPO:P4LT:YVYX:2BKW:2DBI:TV3Y:3ZRM:FOET
WARNING: No swap limit support
```

**uname -a:**

```
Linux vagrant-ubuntu-trusty-64 3.13.0-67-generic #110-Ubuntu SMP Fri Oct 23 13:24:41 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Environment details:**
Virtualbox

**How reproducible:**
Always

**Steps to Reproduce:**
```
mkdir bar
cat > Dockerfile <<EOF
FROM busybox
COPY /bar /foo/bar
RUN ls -ln /
RUN ls /foo
EOF
docker build .

```

**Actual Results:**
```
Sending build context to Docker daemon 3.584 kB
Step 1 : FROM busybox
 ---> 3240943c9ea3
Step 2 : COPY /bar /foo/bar
 ---> ce5fad4f8fb4
Removing intermediate container 0eaaf311eb12
Step 3 : RUN ls -ln /
 ---> Running in aeef126fbf91
total 40
drwxr-xr-x    2 0        0            12288 Feb 16 22:22 bin
drwxr-xr-x    5 0        0              360 Feb 17 13:10 dev
drwxr-xr-x    2 0        0             4096 Feb 17 13:10 etc
drw-------    3 65534    65534         4096 Feb 17 13:10 foo
drwxr-xr-x    2 99       99            4096 Feb 16 22:22 home
dr-xr-xr-x  137 65534    65534            0 Feb 17 13:10 proc
drwxr-xr-x    2 0        0             4096 Feb 16 22:22 root
dr-xr-xr-x   13 65534    65534            0 Feb 17 13:10 sys
drwxrwxrwt    2 0        0             4096 Feb 16 22:22 tmp
drwxr-xr-x    3 0        0             4096 Feb 16 22:22 usr
drwxr-xr-x    4 0        0             4096 Feb 16 22:22 var
 ---> 68f383a45bba
Removing intermediate container aeef126fbf91
Step 4 : RUN ls /foo
 ---> Running in 78415e2c91f3
ls: can't open '/foo': Permission denied
The command '/bin/sh -c ls /foo' returned a non-zero code: 1
```

**Expected Results:**
The Dockerfile building successfully.

**Additional info:**
Notice the owner of the /foo directory that can be seen in the actual result, it is 65534 instead of 0.
If the destination directory (/foo) exists before the copy then the content to be copied (/foo/bar) will be copied correctly and given the right owner.

**Comment by cpuguy83** [comment]
ping @estesp

**Comment by estesp** [comment]
Hmm--I thought this was fixed prior to 1.9; will need to add some tests around this area, but also fix the issue as well.

**Comment by thaJeztah** [comment]
Thanks @estesp, I'll assign you to this then :innocent:
